timemachine/fe/dummy.py

```diff
@@ -0,0 +1,631 @@
+    Generate all choices for valid anchor groups. An anchor group
+    is an ordered sequence of core atoms starting with the anchor (a-b-c) that
+    are connected by bonds. In other words, an anchor group is a rooted subtree
+    with three nodes spanning from the anchor.
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        rdkit molecule
+
+    core: list or set or iterable
+        core atoms
+
+    anchor: int
+        atom we're initializing the search over
+
+    Returns
+    -------
+    3-tuple
+        Returns anchor groups of size 1, 2, and 3. Size 1 group will always
+        be [[anchor]]. Size 2 groups will be [[anchor, x], ...] and size 3
+        groups will be [[anchor,x,y,], ...]
+
+    """
+
+    assert anchor in core
+
+    # We perform a bfs of depth 3 starting from the anchor.
+    layer_1_groups = [[anchor]]
+    layer_2_groups = _add_successors(mol, core, layer_1_groups)
+    layer_3_groups = _add_successors(mol, core, layer_2_groups)
+
+    # sanity assertions to make sure we don't have duplicate idxs
+    for g in layer_2_groups:
+        assert len(set(g)) == len(g)
+
+    for g in layer_3_groups:
+        assert len(set(g)) == len(g)
+
+    return layer_1_groups, layer_2_groups, layer_3_groups
+
+
+def identify_root_anchors(mol, core, dummy_atom):
+    """
+    Identify the root anchor(s) for a given atom. A root anchor is defined as the starting atom
+    in an anchor group (comprised of up to 3 anchor atoms). If this returns multiple root anchors
+    then the dummy_atom is a bridge between multiple core atoms.
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        rdkit molecule
+
+    core: list or set or iterable
+        core atoms
+
+    dummy_atom: int
+        atom we're initializing the search over
+
+    Returns
+    -------
+    list of int
+        List of root anchors that the dummy atom is connected to.
+
+    """
+
+    assert len(set(core)) == len(core)
+
+    core = set(core)
+
+    assert dummy_atom not in core
+
+    # first convert to a dense graph
+    N = mol.GetNumAtoms()
+    dense_graph = np.zeros((N, N), dtype=np.int32)
+
+    for bond in mol.GetBonds():
+        i, j = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
+        dense_graph[i, j] = 1
+        dense_graph[j, i] = 1
+
+    # sparsify to simplify and speed up traversal code
+    sparse_graph = []
+    for row in dense_graph:
+        nbs = []
+        for col_idx, col in enumerate(row):
+            if col == 1:
+                nbs.append(col_idx)
+        sparse_graph.append(nbs)
+
+    def dfs(i, visited):
+        if i in visited:
+            return
+        else:
+            visited.add(i)
+            if i not in core:
+                for nb in sparse_graph[i]:
+                    dfs(nb, visited)
+            else:
+                return
+
+    visited = set()
+
+    dfs(dummy_atom, visited)
+    anchors = []
+
+    for a_idx in visited:
+        if a_idx in core:
+            anchors.append(a_idx)
```

Comment:
nit: cleaner as list comprehension:
```suggestion
    anchors = [a_idx for a_idx in visited if a_idx in core]
```

Reply:
thanks, i still suck at list comprehensions

Reply:
I'm going to just add these in manually as opposed to accepting the commit suggestion to avoid having to deal with conflicts since the code has updated locally on my end.